src/main/java/com/bazel_diff/BazelSourceFileTarget.java

```diff
@@ -20,24 +20,24 @@
 
     BazelSourceFileTargetImpl(String name, byte[] digest, Path workingDirectory) throws IOException, NoSuchAlgorithmException {
         this.name = name;
-        byte[] data = null;
+        MessageDigest finalDigest = MessageDigest.getInstance("SHA-256");
         if (workingDirectory != null && name.startsWith("//")) {
             String filenameSubstring = name.substring(2);
             String filenamePath = filenameSubstring.replaceFirst(":", "/");
-            File sourceFile = new File(workingDirectory.toString(), filenamePath);
-            if (sourceFile.isFile() && sourceFile.canRead()) {
-                ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
-                outputStream.write(Files.readAllBytes(sourceFile.toPath()));
-                outputStream.write(digest);
-                data = outputStream.toByteArray();
-                outputStream.close();
-            }
-        } else {
-            data = digest;
-        }
-        MessageDigest finalDigest = MessageDigest.getInstance("SHA-256");
-        if (data != null) {
-            finalDigest.update(data);
+            Path absoluteFilePath = Paths.get(workingDirectory.toString(), filenamePath);
+            try (RandomAccessFile sourceFile = new RandomAccessFile(absoluteFilePath.toString(), "r")) {
```

Comment:
Note: Using a [buffered stream](https://docs.oracle.com/javase/7/docs/api/java/io/BufferedInputStream.html) would be another option to efficiently calculate the digest. Example of this can be found in apache commons - https://github.com/apache/commons-codec/blob/master/src/main/java/org/apache/commons/codec/digest/DigestUtils.java#L1417

Reply:
That's a bit slow as well, it still takes ~200 seconds. Probably we need to have different handling for small file and large file. We can read file chunk by chunk only if it's larger than 1 Mb.